### PR TITLE
Assert valid composing TextRange

### LINKED ISSUE #49989: RangeError: Value not in range
## Steps to Reproduce

I got this crash on crash collect server and I can't reproduce this on my iPhone.
through the server logs we can see，when this crash occur , the next crash log is 

```bash
NoSuchMethodError: The method 'findRenderObject' was called on null.\
Receiver: null\
Tried calling: findRenderObject()\
#0 EditableTextState._onCursorColorTick (package:flutter/src/widgets/editable_text.dart:1347)\
#1_AnimationController&Animation&AnimationEagerListenerMixin&AnimationLocalListenersMixin.notifyListeners (package:flutter/src/animation/listener_helpers.dart:126)\
#2 AnimationController.set:value (package:flutter/src/animation/animation_controller.dart:351)\
#3 EditableTextState._startCursorTimer (package:flutter/src/widgets/editable_text.dart:1274)\
```

```bash
RangeError: Value not in range: 8\
#0 _StringBase.substring (dart:core-patch/string_patch.dart:386)\
#1 TextRange.textInside (package:flutter/src/services/text_editing.dart:72)\
#2 EditableTextState.buildTextSpan (package:flutter/src/widgets/editable_text.dart:1475)\
#3 EditableTextState.build.<anonymous closure> (package:flutter/src/widgets/editable_text.dart:1422)\
#4 ScrollableState.build (package:flutter/src/widgets/scrollable.dart:584)\
#5 StatefulElement.build (package:flutter/src/widgets/framework.dart:3825)\
#6 ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3739)\
#7 Element.rebuild (package:flutter/src/widgets/framework.dart:3565)\
#8 StatefulElement.update (package:flutter/src/widgets/framework.dart:3894)\
#9 Element.updateChild (package:flutter/src/widgets/framework.dart:2753)\
#10 ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3750)\
#11 Element.rebuild (package:flutter/src/widgets/framework.dart:3565)\
#12 BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2278)\
#13 _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:700)\
#14 _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:286)\
#15 _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1012)\
#16 _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:952)\
#17 _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:864)\
#18 FrameMetrics._handleDrawFrame (package:mt_flutter_route/monitor/fps/frame_metrics.dart:42)\
#19 FrameMetrics._handleDrawFrame (package:flutter_metrics_plugin/metrics/frame_metrics.dart:42)\
#20 _rootRun (dart:async/zone.dart:1124)\
#21 _CustomZone.run (dart:async/zone.dart:1021)\
#22 _CustomZone.runGuarded (dart:async/zone.dart:923)\
#23 _invoke (dart:ui/hooks.dart:219)\
#24 _drawFrame (dart:ui/hooks.dart:178)\
```

```console
[✓] Flutter (Channel v1.5.4-hotfixes, v1.5.4-hotfix.2-pre.1, on
    Mac OS X 10.14.6 18G87, locale zh-Hans-CN)
    • Flutter version 1.5.4-hotfix.2-pre.1 at
      /Users/yantao/flutter
    • Framework revision 953bbe2ccd (9 months ago), 2019-05-02
      09:59:16 -0700
    • Engine revision 52c7a1e849
    • Dart version 2.3.0 (build 2.3.0-dev.0.5 a1668566e5)
```

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

- Try to catch invalid composing ranges early in `TextEditingController`s, rather than during build time.
- Ignore the composing range if it's invalid during build.

## Related Issues

fixes https://github.com/flutter/flutter/issues/49989
related: https://github.com/flutter/flutter/issues/33730

## Tests

I added the following tests:

- TextEditingValue.isComposingRangeValid
- Asserts if composing text is not valid

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -134,7 +134,13 @@ class TextEditingController extends ValueNotifier<TextEditingValue> {
   /// This constructor treats a null [value] argument as if it were
   /// [TextEditingValue.empty].
   TextEditingController.fromValue(TextEditingValue value)
-    : super(value ?? TextEditingValue.empty);
+    : assert(
+        value == null || !value.composing.isValid || value.isComposingRangeValid,
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:138]: It's somewhat confusing that `value.composing.isValid` and `value.isComposingRangeValid` seem so similar but mean different things.  I can't think of a way to make it more clear though 🤷

**Review by justinmc — APPROVED** [review]
LGTM 

+1 for making problems like this more clear to users 👍

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:138]: I guess we could use null to represent an editing value that has no composing range. That would be a breaking change.

Comment (justinmc) [reply packages/flutter/lib/src/widgets/editable_text.dart:138]: Probably not worth it.

**Pull request merged**